src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+import Chainweb.Logger (Logger, addLabel, logFunctionText)
+import Chainweb.Pact.Backend.Compaction (TargetBlockHeight(..))
+import Chainweb.Pact.Backend.Compaction qualified as C
+import Chainweb.Pact.Backend.PactState (PactRow(..), getLatestPactStateUpperBound', PactRowContents(..))
+import Chainweb.Pact.Backend.PactState.EmbeddedHashes (EncodedSnapshot(..), grands)
+import Chainweb.Pact.Backend.Types (SQLiteEnv(..))
+import Chainweb.Pact.Backend.Utils (withSqliteDb)
+import Chainweb.Storage.Table.RocksDB (RocksDb, withReadOnlyRocksDb, modernDefaultOptions)
+import Chainweb.TreeDB (seekAncestor)
+import Chainweb.Utils (fromText, toText, sshow)
+import Chainweb.Version (ChainwebVersion(..), ChainwebVersionName)
+import Chainweb.Version.Mainnet (mainnet)
+import Chainweb.Version.Registry (lookupVersionByName)
+import Chainweb.Version.Utils (chainIdsAt)
+import Chainweb.WebBlockHeaderDB (initWebBlockHeaderDb)
+import Chainweb.WebBlockHeaderDB (webBlockHeaderDb)
+import Control.Applicative ((<|>), many)
+import Control.Applicative (optional)
+import Control.Lens ((^.))
+import Control.Monad (forM, forM_, when)
+import Crypto.Hash (hashWith, hashInitWith, hashUpdate, hashFinalize)
+import Crypto.Hash.Algorithms (SHA3_256(..))
+import Data.ByteArray qualified as Memory
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.ByteString.Base16 qualified as Base16
+import Data.ByteString.Builder qualified as BB
+import Data.ByteString.Lazy qualified as BL
+import Data.ByteString.Lazy.Char8 qualified as BLC8
+import Data.Foldable qualified as F
+import Data.HashMap.Strict qualified as HM
+import Data.IORef
+import Data.Int (Int64)
+import Data.List qualified as List
+import Data.Map (Map)
+import Data.Map.Merge.Strict qualified as Merge
+import Data.Map.Strict qualified as Map
+import Data.Maybe (catMaybes)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Encoding qualified as Text
+import Data.Vector (Vector)
+import Data.Vector qualified as Vector
+import Data.Word (Word64)
+import Database.SQLite3.Direct (Database)
+import GHC.Stack (HasCallStack)
+import Options.Applicative (ParserInfo, Parser, (<**>))
+import Options.Applicative qualified as O
+import Pact.JSON.Encode qualified as J
+import Pact.Types.SQLite (RType(..), SType(..))
+import Pact.Types.SQLite qualified as Pact
+import Patience.Map qualified as P
+import Streaming.Prelude (Stream, Of)
+import Streaming.Prelude qualified as S
+import System.Directory (copyFile, createDirectoryIfMissing, doesFileExist)
+import System.Exit (exitFailure)
+import System.FilePath ((</>))
+import System.LogLevel (LogLevel(..))
+import System.Logger.Types qualified as YAL
+import UnliftIO.Async (pooledMapConcurrentlyN_)
+
+-- | Compute the "Grand Hash" of a given chain.
+computeGrandHash :: Database -> BlockHeight -> IO ByteString
```

Comment:
Lars has many questions about this algorithm. In Haskell, it is very simple to represent. 
We may want to compute the hash incrementally either in SQL or some persisted data structure (persisted to SQL or something else).
This algorithm is correct as per the spec that compaction first designed. But it may need to change. It's not even clear if we're going to continue using SHA3_256!

@larskuhtz if you have time to review anything in this PR, it should be `computeGrandHash`, and its transitive function calls that concern hashing/pact state.